one-ui/src/main/java/com/marklogic/hub/oneui/controllers/EnvironmentController.java

```diff
@@ -90,31 +96,109 @@ public JsonNode getProject() {
 
     @RequestMapping(value = "/install", method = RequestMethod.POST)
     @ResponseBody
-    public JsonNode install(@RequestBody ObjectNode payload) throws Exception {
-        String originalDirectory = environmentService.getProjectDirectory();
+    public void install(@RequestBody ObjectNode payload) throws Exception {
+        String directory = payload.get("directory").asText("");
+
         final Exception[] dataHubConfigurationException = {null};
+        HubDeployStatusListener listener = getInstallListener(dataHubConfigurationException, false);
+
+        install(directory, listener, dataHubConfigurationException);
+    }
+
+    @RequestMapping(value = "/project-download", produces = "application/zip")
+    public void downloadProject(HttpServletRequest request, HttpServletResponse response) {
+        HubProject project = hubConfig.getHubProject();
+        response.setContentType(MediaType.APPLICATION_OCTET_STREAM_VALUE);
+        response.addHeader("Content-Disposition", "attachment; filename=datahub-project.zip");
+        try (OutputStream out = response.getOutputStream()) {
+            project.exportProject(out);
+            response.flushBuffer();
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to download project;cause: " + e.getMessage());
+        }
+    }
+
+    @RequestMapping(value = "/reset", method = RequestMethod.POST)
+    @ResponseBody
+    public JsonNode reset() {
+        environmentService.reset();
+        ObjectNode obj = mapper.createObjectNode();
+        obj.put("reset", true);
+        return obj;
+    }
+
+    @RequestMapping(value = "/project-info", method = RequestMethod.GET)
+    @ResponseBody
+    public JsonNode getProjectInfo() {
+        return environmentService.getProjectInfo();
+    }
+
+    @RequestMapping(value = "/project-upload", method = RequestMethod.POST)
+    @ResponseBody
+    public void uploadProject(@RequestParam("zipfile") MultipartFile uploadedFile) throws Exception {
+        final Exception[] dataHubConfigurationException = {null};
+        HubDeployStatusListener listener = getInstallListener(dataHubConfigurationException, false);
+
+        HubProject project = hubConfig.getHubProject();
+
+        //backup first
+        backupExistingFSProject(project, listener);
+
+        // delete contents from the current project folder
+        cleanExistingFSProject(project, listener);
+
+        // extract the uploaded & zipped project file into the current project folder
+        extractZipProject(project, uploadedFile, listener);
+
+        HubDeployStatusListener uninstalllistener = getInstallListener(dataHubConfigurationException, true);
+        dataHubService.unInstall(uninstalllistener);
+
+        install(project.getProjectDirString(), listener, dataHubConfigurationException);
+    }
+
+    private HubDeployStatusListener getInstallListener(final Exception[] dataHubConfigurationException, boolean isUninstall) {
```

Comment:
I am generally not big on anonymous classes like this. I find them more difficult to read and understand, and they can also complicate the dependencies on a class. For example, EnvironmentController has a dependency on SimpMessagingTemplate, but it doesn't need it - this anonymous impl needs it.

Let's move this into a real class - call it "UIDeployListener" or something like that. Move the SimpMessagingTemplate dependency to it, which can be passed in via the constructor, along with "boolean isUninstall".

Reply:
I can create a UIDeployListener class.